src/twisted/web/test/test_websocket.py

```diff
@@ -0,0 +1,344 @@
+from twisted.python.failure import Failure
+from twisted.test.iosim import ConnectionCompleter, IOPump
+from twisted.trial.unittest import SynchronousTestCase
+from twisted.web._responses import BAD_REQUEST
+from twisted.web.client import Agent, readBody
+from twisted.web.iweb import IRequest
+from twisted.web.resource import Resource
+from twisted.web.server import NOT_DONE_YET, Request, Site
+from twisted.web.static import Data
+
+WSP = TypeVar("WSP", bound="WebSocketProtocol")
+shouldSkip = False
+
+
+class WeirdResource(Resource):
+    def render_GET(self, request: Request) -> bytes:
+        """
+        Per U{the wsproto documentation
+        <https://python-hyper.org/projects/wsproto/en/latest/api.html#wsproto.events.RejectConnection.has_body>}:
+
+            - The only scenario in which the caller receives a RejectConnection
+              with C{has_body == False} is if the peer violates sends an
+              informational status code (1xx) other than 101
+
+        This is a weird edge case so we provoke it.
+        """
+        request.setResponseCode(102)
+        return b""
+
+
+class SlowResource(Resource):
+    def render_GET(self, request: Request) -> int:
+        self.request = request
+        return NOT_DONE_YET
+
+
+try:
+    __import__("wsproto")
+except ImportError:
+    shouldSkip = True
+else:
+    from twisted.web.websocket import (
+        ConnectionRejected,
+        WebSocketClientEndpoint,
+        WebSocketClientFactory,
+        WebSocketProtocol,
+        WebSocketResource,
+        WebSocketServerFactory,
+        WebSocketTransport,
+    )
+
+    @dataclass
+    class MyWSP(WebSocketProtocol):
+        pongs: list[bytes] = field(default_factory=list)
+        wasLost: Failure | None = None
+
+        def negotiationStarted(self, transport: WebSocketTransport) -> None:
+            self.transport = transport
+
+        def negotiationFinished(self) -> None:
+            ...
+
+        def connectionLost(self, reason: Failure) -> None:
+            self.wasLost = reason
+
+        def bytesMessageReceived(self, data: bytes) -> None:
+            if data == b"request":
+                self.transport.sendBytesMessage(b"\x00resp\x01onse\xff")
+            else:
+                self.bDeferred.callback(data)
+
+        def textMessageReceived(self, data: str) -> None:
+            if data == "request":
+                self.transport.sendTextMessage("response")
+            else:
+                self.deferred.callback(data)
+
+        def pongReceived(self, payload: bytes) -> None:
+            self.pongs.append(payload)
+
+        def sendRequest(self) -> Deferred[str]:
+            """
+            Send a text message to the server and expect a response.
+            """
+            self.deferred: Deferred[str] = Deferred()
+            self.transport.sendTextMessage("request")
+            return self.deferred
+
+        def bytesRequest(self) -> Deferred[bytes]:
```

Comment:
similar to make it clear that this is testing only method

```suggestion
        def pumpBytesRequest(self) -> Deferred[bytes]:
```

Reply:
This doesn't issue any commands to the pump though? It just sends a bytes request? The fact that it's a method on a testing fixture seems like a strong hint :)